src/decorators/headerFooter.js

```diff
@@ -9,16 +9,30 @@ import Template from '../components/Template.js';
 
 function headerFooter(ComposedComponent) {
   class HeaderFooter extends React.Component {
-    componentWillMount() {
-      // Only add header/footer if a template is defined
-      this._header = this.getTemplate('header');
-      this._footer = this.getTemplate('footer');
-      this._classNames = {
-        root: cx(this.props.cssClasses.root),
-        body: cx(this.props.cssClasses.body)
+    constructor(props) {
+      super(props);
+      this.handleHeaderClick = this.handleHeaderClick.bind(this);
+      this.state = {
+        collapsed: props.collapsed
       };
+
+      this._headerElement = this._getElement({
+        type: 'header',
+        handleClick: props.collapsable === true ? this.handleHeaderClick : null
+      });
+
+      this._cssClasses = {
+        root: cx('ais-root', this.props.cssClasses.root),
+        body: cx('ais-body', this.props.cssClasses.body)
+      };
+
+      this._footerElement = this._getElement({type: 'footer'});
     }
-    getTemplate(type) {
+    shouldComponentUpdate(nextProps, nextState) {
+      return nextState.collapsed === false ||
+        nextState !== this.state;
+    }
+    _getElement({type, handleClick = null}) {
```

Comment:
> What about _getComponent since we get a react component?

Oh yeah, this fine grained difference is insane.. https://facebook.github.io/react/blog/2015/12/18/react-components-elements-and-instances.html